File: leetcode-algorithms/093. Restore IP Addresses/Solution.cpp
// Given a string containing only digits, restore it by returning all possible valid IP address combinations.
//
// Example:
//
// Input: "25525511135"
// Output: ["255.255.11.135", "255.255.111.35"]
class Solution {
public:
    vector<string> restoreIpAddresses(string s) {
        vector<string> rslt;
        vector<vector<string>> rsltv;
        vector<string> temp;
        travel(s, 0, temp, rsltv);

        for (auto v : rsltv) {
            string s = "";
            for (int i = 0; i < v.size(); i++) {
                s += v[i];
                if (i == v.size() - 1) break;
                s += '.';
            }
            rslt.push_back(s);
        }
        return rslt;
    }

    void travel(string& s, int pos, vector<string>& temp, vector<vector<string>>& rslt) {
        if (pos == s.size()) {
            if (temp.size() == 4) {
                rslt.push_back(temp);
                return;
            }
        } else if (pos > s.size()) return;
        else if (temp.size() > 4) return;

        if (s[pos] == '0') {
            temp.push_back("0");
            travel(s, pos + 1, temp, rslt);
            temp.pop_back();
        } else {
            for (int i = pos; i < min(pos + 4, int(s.size())); i++) {
                string sub = s.substr(pos, i - pos + 1);
                if (stoi(sub) <= 255) {
                    temp.push_back(sub);
                    travel(s, i + 1, temp, rslt);
                    temp.pop_back();
                }
            }
        }
    }
};

File: leetcode-algorithms/093. Restore IP Addresses/Solution.java
class Solution {
    public List<String> restoreIpAddresses(String s) {
        List<String> temp = new ArrayList<>();
        impl(s, 0, temp);

        List<String> f = new ArrayList<>();
        for (List<String> li : rslt) {
            String t = "";
            for (String str: li) {
                t += str;
                t += ".";
            }

            f.add(t.substring(0, t.length() - 1));
        }

        return f;
    }

    void impl(String s, int pos, List<String> temp) {
        if ((temp.size() == 4) && (pos != s.length())) return;

        if ((pos == s.length()) && (temp.size() == 4)) {
            List<String> t = new ArrayList<>(temp);
            rslt.add(t);
            return;
        } else if (pos >= s.length()) return;

        if (s.charAt(pos) == '0') {
            temp.add("0");
            impl(s, pos + 1, temp);
            temp.remove(temp.size() - 1);
        } else {
            for (int i = pos; i < Math.min(pos + 3, s.length()); i++) {
                String sub = s.substring(pos, i + 1);
                if (Integer.valueOf(sub) <= 255) {
                    temp.add(sub);

                    impl(s, i + 1, temp);
                    temp.remove(temp.size() - 1);
                }
            }
        }
    }

    List<List<String>> rslt = new ArrayList<>();
}
